File: OtherLanguages/SnakeGame.lua
-- main.lua
-- Author: Sanne Karibo
-- Snake Game in Lua using Love2D

local width, height = 64, 48
local cellSize = 10
local maxLength = 100

local snake = {}
local dir = "right"
local food = {}
local score = 0
local gameOver = false

function love.load()
    love.window.setMode(width * cellSize, height * cellSize)
    -- Initialize snake
    for i = 5, 1, -1 do
        table.insert(snake, {x = 20 - i, y = 10})
    end
    generateFood()
end

function generateFood()
    while true do
        food.x = math.random(0, width - 1)
        food.y = math.random(0, height - 1)
        local collision = false
        for _, segment in ipairs(snake) do
            if segment.x == food.x and segment.y == food.y then
                collision = true
                break
            end
        end
        if not collision then break end
    end
end

function love.update(dt)
    if gameOver then return end

    -- Move snake every 0.15 seconds
    if not love.timer then return end
    if not love.updateTimer then love.updateTimer = 0 end
    love.updateTimer = love.updateTimer + dt
    if love.updateTimer < 0.15 then return end
    love.updateTimer = 0

    local head = snake[1]
    local newHead = {x = head.x, y = head.y}

    if dir == "right" then newHead.x = newHead.x + 1
    elseif dir == "left" then newHead.x = newHead.x - 1
    elseif dir == "up" then newHead.y = newHead.y - 1
    elseif dir == "down" then newHead.y = newHead.y + 1
    end

    -- Check collisions
    if newHead.x < 0 or newHead.x >= width or newHead.y < 0 or newHead.y >= height then
        gameOver = true
        return
    end

    for i = 1, #snake do
        if snake[i].x == newHead.x and snake[i].y == newHead.y then
            gameOver = true
            return
        end
    end

    table.insert(snake, 1, newHead)

    if newHead.x == food.x and newHead.y == food.y then
        score = score + 10
        generateFood()
        if #snake > maxLength then
            table.remove(snake)
        end
    else
        table.remove(snake)
    end
end

function love.draw()
    love.graphics.clear(0, 0, 0)

    -- Draw snake
    for i, segment in ipairs(snake) do
        if i == 1 then
            love.graphics.setColor(1, 1, 0) -- yellow head
        else
            love.graphics.setColor(0, 1, 0) -- green body
        end
        love.graphics.rectangle("fill", segment.x * cellSize, segment.y * cellSize, cellSize - 1, cellSize - 1)
    end

    -- Draw food
    love.graphics.setColor(1, 0, 0)
    love.graphics.circle("fill", food.x * cellSize + cellSize / 2, food.y * cellSize + cellSize / 2, cellSize / 2 - 1)

    -- Draw score
    love.graphics.setColor(1, 1, 1)
    love.graphics.print("Score: " .. score, 10, 10)

    if gameOver then
        love.graphics.printf("Game Over! Final Score: " .. score, 0, love.graphics.getHeight() / 2, love.graphics.getWidth(), "center")
    end
end

function love.keypressed(key)
    if key == "right" and dir ~= "left" then dir = "right"
    elseif key == "left" and dir ~= "right" then dir = "left"
    elseif key == "up" and dir ~= "down" then dir = "up"
    elseif key == "down" and dir ~= "up" then dir = "down"
    elseif key == "escape" then love.event.quit() end
end

File: OtherLanguages/SnakeGame.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.Random;

public class SnakeGame extends JPanel implements ActionListener, KeyListener {

    // Constants and game variables
    private static final int WIDTH = 640;
    private static final int HEIGHT = 480;
    private static final int DOT_SIZE = 10;
    private static final int MAX_DOTS = 100;
    private static final int ALL_DOTS = (WIDTH / DOT_SIZE) * (HEIGHT / DOT_SIZE);

    private final int[] x = new int[MAX_DOTS];
    private final int[] y = new int[MAX_DOTS];
    private int length;
    private int dir; // 1=Right, 2=Down, 3=Left, 4=Up
    private int foodX;
    private int foodY;
    private boolean gameOver;
    private int score;

    private Timer timer;
    private Random random;

    // Constructor to setup game
    public SnakeGame() {
        setPreferredSize(new Dimension(WIDTH, HEIGHT));
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);

        random = new Random();
        initGame();
    }

    private void initGame() {
        length = 5;
        dir = 1; // start moving right
        score = 0;
        gameOver = false;

        // Initialize snake position (centered horizontally)
        for (int i = 0; i < length; i++) {
            x[i] = 200 - i * DOT_SIZE;
            y[i] = 100;
        }

        locateFood();

        timer = new Timer(100, this); // update every 100 ms
        timer.start();
    }

    private void locateFood() {
        foodX = random.nextInt(WIDTH / DOT_SIZE) * DOT_SIZE;
        foodY = random.nextInt(HEIGHT / DOT_SIZE) * DOT_SIZE;
    }

    @Override
    public void paintComponent(Graphics g) {
        super.paintComponent(g);

        if (!gameOver) {
            drawScore(g);
            drawFood(g);
            drawSnake(g);
        } else {
            gameOverScreen(g);
        }
    }

    private void drawScore(Graphics g) {
        g.setColor(Color.CYAN);
        g.setFont(new Font("Arial", Font.BOLD, 14));
        g.drawString("Score: " + score, 10, 20);
    }

    private void drawFood(Graphics g) {
        g.setColor(Color.RED);
        g.fillOval(foodX, foodY, DOT_SIZE, DOT_SIZE);
    }

    private void drawSnake(Graphics g) {
        // Head in bright yellow
        g.setColor(Color.YELLOW);
        g.fillRect(x[0], y[0], DOT_SIZE, DOT_SIZE);

        // Body segments in shades of green
        for (int i = 1; i < length; i++) {
            int greenValue = Math.max(50, 255 - (i * 10));
            g.setColor(new Color(0, greenValue, 0));
            g.fillRect(x[i], y[i], DOT_SIZE, DOT_SIZE);
        }
    }

    private void move() {
        // Move body segments
        for (int i = length - 1; i > 0; i--) {
            x[i] = x[i - 1];
            y[i] = y[i - 1];
        }

        // Move head based on direction
        switch (dir) {
            case 1: x[0] += DOT_SIZE; break; // right
            case 2: y[0] += DOT_SIZE; break; // down
            case 3: x[0] -= DOT_SIZE; break; // left
            case 4: y[0] -= DOT_SIZE; break; // up
        }
    }

    private void checkFood() {
        if (x[0] == foodX && y[0] == foodY) {
            if (length < MAX_DOTS) {
                length++;
                x[length - 1] = x[length - 2];
                y[length - 1] = y[length - 2];
            }
            score += 10;
            Toolkit.getDefaultToolkit().beep(); // simple sound effect
            locateFood();
        }
    }

    private void checkCollision() {
        // Check collision with walls
        if (x[0] < 0 || x[0] >= WIDTH || y[0] < 0 || y[0] >= HEIGHT) {
            gameOver = true;
            Toolkit.getDefaultToolkit().beep();
            timer.stop();
        }

        // Check collision with itself
        for (int i = 1; i < length; i++) {
            if (x[0] == x[i] && y[0] == y[i]) {
                gameOver = true;
                Toolkit.getDefaultToolkit().beep();
                timer.stop();
            }
        }
    }

    private void gameOverScreen(Graphics g) {
        String msg1 = "Game Over!";
        String msg2 = "Final Score: " + score;
        String msg3 = "Press any key to exit...";

        g.setColor(Color.RED);
        g.setFont(new Font("Arial", Font.BOLD, 40));
        FontMetrics fm = getFontMetrics(g.getFont());
        g.drawString(msg1, (WIDTH - fm.stringWidth(msg1)) / 2, HEIGHT / 2 - 50);

        g.setFont(new Font("Arial", Font.BOLD, 24));
        fm = getFontMetrics(g.getFont());
        g.drawString(msg2, (WIDTH - fm.stringWidth(msg2)) / 2, HEIGHT / 2);

        g.drawString(msg3, (WIDTH - fm.stringWidth(msg3)) / 2, HEIGHT / 2 + 50);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        if (!gameOver) {
            move();
            checkFood();
            checkCollision();
        }
        repaint();
    }

    // Handle key input to change direction
    @Override
    public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();

        switch (key) {
            case KeyEvent.VK_RIGHT:
                if (dir != 3) dir = 1;
                break;
            case KeyEvent.VK_DOWN:
                if (dir != 4) dir = 2;
                break;
            case KeyEvent.VK_LEFT:
                if (dir != 1) dir = 3;
                break;
            case KeyEvent.VK_UP:
                if (dir != 2) dir = 4;
                break;
            default:
                // If game over and any key pressed, exit
                if (gameOver) System.exit(0);
        }
    }

    @Override public void keyReleased(KeyEvent e) {}
    @Override public void keyTyped(KeyEvent e) {}

    // Main method to start the game
    public static void main(String[] args) {
        JFrame frame = new JFrame("Snake Game - Sanne Karibo");
        SnakeGame gamePanel = new SnakeGame();

        frame.add(gamePanel);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.pack();
        frame.setResizable(false);
        frame.setLocationRelativeTo(null);
        frame.setVisible(true);

        // Show prescreen dialog
        JOptionPane.showMessageDialog(frame,
                "Snake Game\nCreated by Sanne Karibo\n\nPress OK to start...",
                "Welcome", JOptionPane.INFORMATION_MESSAGE);
    }
}
